### DISCUSSION TITLE
Counted strings and Unicode

### DISCUSSION THREAD

**Discussion by massung** [ForthHub/discussion#94] [Q&A]
For COUNT, TYPE, and their compatriots, does anyone know if the "count" refers to code-points (read: characters) or bytes? My feeling right now is that it isn't specified and up to an implementation.

Either way, have any of you implemented a Forth that supported Unicode (using c-strings), and did you find one count type over the other had any particular benefits?

**Comment by AntonErtl** [comment]
It refers to chars (aka pchars), i.e., code units in Unicode terminology, and these usually are bytes.  Code points may take more than one char if you have the xchar extension.

Gforth supports Unicode in the form of UTF-8, and you can certainly put UTF-8 in a counted string.  If you write, say c" ..." with some UTF-8 content, the count reflects the number of bytes.  The benefit if this is that you know how much memory you need to MOVE for copying, and that, e.g., COUNT TYPE just works.  The only benefit of the other alternative is that you would know how many code points the string contains, a pretty useless piece of information.

**Comment by mitra42** [comment]
As I implemented webForth, which used eForth as a base implementation, it was pretty clear that people weren't thinking Unicode ! I'm planning to go through and retool it around UTF8 since it seems absurd to be english-only these days.

Biggest problem I found were assumptions that it was one byte per character, which obviously isn't true for Unicode. Its easy to retrofit TYPE so that instead of looping over the string emitting single bytes it passes a UTF8 string to the low level routine. I think it will be harder to work with things like CHAR which presume they can read a character, but return a byte.

**Comment by noqsi** [comment]
[LSE64](https://github.com/noqsi/LSE64) simply has one character per cell, straight Unicode encoding internally. It uses the encoding specified by the locale for character I/O.

John Doty              Noqsi Aerospace, Ltd.
jpd@noqsi.com
